source/_posts/2020-07-28-Kubernetes-controllers-A-New-Hope.md

```diff
@@ -0,0 +1,308 @@
+
+{% note %}
+One of WebAssembly's unique attributes is the ability to run sandboxed without relying on OS process boundaries. Requiring a 1-to-1 correspondence between wasm instances and heavyweight OS processes would take away this key advantage for many use cases. Fork/exec are the obvious example of an API that's difficult to implement well if you don't have POSIX-style processes, but a lot of other things in POSIX are tied to processes too. So it isn't a simple matter to take POSIX, or even a simple subset of it, to WebAssembly.
+{% endnote %}
+
+A socket, for example, is a process wide resource. This means that if module A open a TCP socket, module B might access to it if we don't implement proper security countermeasures. 
+
+### Designing the ABI
+
+Given our knowledge and studies of the current state of art for Wasm ABIs, we found out there are 3 possible (non mutually excluded) approaches to the ABI design:
+
+* **High level ABI**: expose the Kubernetes client functionalities
+* **Medium level ABI**: proxy the HTTP client functionalities (send HTTP request, receive HTTP response)
+* **Low level ABI**: proxy the syscalls (`epoll`, `read`, `write`, `bind`, ...)
+
+We think that a good solution should mix all these 3 _"levels"_. Some low-level APIs are always necessary for basic things like logging, get configuration from environment, set timeout and so on. Proxying HTTP can be useful to invoke services outside Kubernetes (e.g. to trigger a cloud vendor API to enable a service).
+Exposing the Kubernetes client APIs unlocks a great potential to optimize the controllers, to name few: The host needs to watch only one time a `ConfigMap` required by all different controllers only once, a single cache layer can be shared among all controllers watching the same resources, RBAC checks might be done directly on the host side.
+All these optimizations have the potential to greatly reduce network traffic and general load in a Kubernetes cluster, especially as more and more extensions get added to it.
+
+{% note info %}
+Just as an example, imagine a system where all the controllers watch a `ConfigMap` called `config-logging` to configure the logging level. Assuming there are controllers in separate processes, all of them needs to do long polling HTTP requests to Kubernetes in order to check for changes and reconfigure logging. With the solution proposed here, we need only one long polling HTTP request, reducing by a factor 10 the network activity and CPU cycles on that particular watch (both server side and client side).
+{% endnote %}
+
+Following the above approach, host has always the full control of process resources (open files, open sockets, etc), we allow the modules to perform only certain operations and, most important, we open up for **huge optimizations**.
```

Comment:
```suggestion
```suggestion
Exposing the Kubernetes client APIs unlocks a great potential to optimize the controllers beyond the optimized resource usage indicated above. Most prominently: The so called _informer_ infrastructure can be shared between all the controllers. An informer is the part of a controller that listens for events on given resources. It builds up local caches that effectively reflect the state of said resources in-memory.

In our case, the host would setup these informers and watches just **once** per resource. If two controllers want to watch `ConfigMap`s for example, the host would only need to setup one watch and keep one cache, reducing the amount of network traffic due to event delivery and memory consumption due to caching drastically. That effect becomes more and more pronounced as more extensions are being added.
```
```

Reply:
Better if you double check it again